From 0aea67490830a0a93f4c212a3e7a6523b14572c7 Mon Sep 17 00:00:00 2001
From: Bo Gan <ganb@vmware.com>
Date: Tue, 2 Jun 2020 20:02:06 -0700
Subject: [PATCH] x86/vmware: Fix steal time clock under SEV

STA shared memory should be set to decrypted if SEV is active.

Signed-off-by: Bo Gan <ganb@vmware.com>
---
 arch/x86/kernel/cpu/vmware.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/arch/x86/kernel/cpu/vmware.c b/arch/x86/kernel/cpu/vmware.c
index c34d7e5335da..b67e7bee1234 100644
--- a/arch/x86/kernel/cpu/vmware.c
+++ b/arch/x86/kernel/cpu/vmware.c
@@ -297,9 +297,24 @@ static struct notifier_block vmware_pv_reboot_nb = {
 	.notifier_call = vmware_pv_reboot_notify,
 };
 
+static inline void sev_map_percpu_data(void)
+{
+	int cpu;
+
+	if (!cc_platform_has(CC_ATTR_GUEST_MEM_ENCRYPT))
+		return;
+
+	for_each_possible_cpu(cpu) {
+		early_set_memory_decrypted(
+			(unsigned long)&per_cpu(vmw_steal_time, cpu),
+			sizeof(vmw_steal_time));
+	}
+}
+
 #ifdef CONFIG_SMP
 static void __init vmware_smp_prepare_boot_cpu(void)
 {
+	sev_map_percpu_data();
 	vmware_guest_cpu_init();
 	native_smp_prepare_boot_cpu();
 }
-- 
2.25.1

